File: project/jimmer-sql/src/main/java/org/babyfish/jimmer/sql/ast/impl/ConcatExpression.java
package org.babyfish.jimmer.sql.ast.impl;

import org.babyfish.jimmer.sql.ast.Expression;
import org.babyfish.jimmer.sql.ast.impl.render.AbstractSqlBuilder;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Objects;

class ConcatExpression
        extends AbstractExpression<String>
        implements StringExpressionImplementor {

    private Expression<String> first;

    private List<Expression<String>> others;

    ConcatExpression(Expression<String> first, List<Expression<String>> others) {
        this.first = first;
        this.others = others;
    }

    @Override
    public void accept(@NotNull AstVisitor visitor) {
        ((Ast) first).accept(visitor);
        for (Expression<?> other : others) {
            ((Ast) other).accept(visitor);
        }
    }

    @Override
    public void renderTo(@NotNull AbstractSqlBuilder<?> builder) {
        builder.sql("concat(");
        renderChild((Ast) first, builder);
        for (Expression<?> other : others) {
            builder.sql(", ");
            renderChild((Ast) other, builder);
        }
        builder.sql(")");
    }

    @Override
    protected boolean determineHasVirtualPredicate() {
        return hasVirtualPredicate(first) || hasVirtualPredicate(others);
    }

    @Override
    protected Ast onResolveVirtualPredicate(AstContext ctx) {
        this.first = ctx.resolveVirtualPredicate(first);
        this.others = ctx.resolveVirtualPredicates(others);
        return this;
    }

    @Override
    public int precedence() {
        return 0;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ConcatExpression that = (ConcatExpression) o;
        return first.equals(that.first) && others.equals(that.others);
    }

    @Override
    public int hashCode() {
        return Objects.hash(first, others);
    }
}

File: project/jimmer-sql-kotlin/src/main/kotlin/org/babyfish/jimmer/sql/kt/ast/expression/impl/ConcatExpression.kt
package org.babyfish.jimmer.sql.kt.ast.expression.impl

import org.babyfish.jimmer.sql.ast.impl.Ast
import org.babyfish.jimmer.sql.ast.impl.AstContext
import org.babyfish.jimmer.sql.ast.impl.AstVisitor
import org.babyfish.jimmer.sql.ast.impl.render.AbstractSqlBuilder
import org.babyfish.jimmer.sql.kt.ast.expression.KExpression
import org.babyfish.jimmer.sql.kt.ast.expression.KNonNullExpression
import org.babyfish.jimmer.sql.kt.ast.expression.KNullableExpression

internal open class ConcatExpression(
    private var expressions: List<KExpression<String>>
) : AbstractKExpression<String>() {

    init {
        if (expressions.isEmpty()) {
            throw IllegalArgumentException("No expressions")
        }
    }

    override fun getType(): Class<String> = String::class.java

    override fun precedence(): Int = 0

    override fun accept(visitor: AstVisitor) {
        for (expression in expressions) {
            (expression as Ast).accept(visitor)
        }
    }

    override fun renderTo(builder: AbstractSqlBuilder<*>) {
        builder.sql("concat(")
        var sp = ""
        for (expression in expressions) {
            builder.sql(sp)
            sp = ", "
            (expression as Ast).renderTo(builder)
        }
        builder.sql(")")
    }

    override fun determineHasVirtualPredicate(): Boolean =
        hasVirtualPredicate(expressions)

    override fun onResolveVirtualPredicate(ctx: AstContext): Ast {
        expressions = ctx.resolveVirtualPredicates(expressions)
        return this
    }

    class NonNull(
        expressions: List<KNonNullExpression<String>>
    ) : ConcatExpression(expressions), KNonNullExpression<String>

    class Nullable(
        expressions: List<KExpression<String>>
    ) : ConcatExpression(expressions), KNullableExpression<String>
}